Provide BC typehinting for TableSchema.

dereuromark opened the pull request:
When running migrations, or running tests on bake plugin right now:
```
1) Bake\Test\TestCase\Shell\Task\FixtureTaskTest::testImportOptionsAlternateConnection
TypeError: Argument 1 passed to Bake\Shell\Task\FixtureTask::_generateRecords() must be an instance of
Cake\Database\Schema\Table, instance of Cake\Database\Schema\TableSchema given, called in
D:\...\Apps\sandbox.local\vendor\cakephp\bake\src\Shell\Task\FixtureTask.php on line 194

D:\...\Apps\sandbox.local\vendor\cakephp\bake\src\Shell\Task\FixtureTask.php:322
D:\...\Apps\sandbox.local\vendor\cakephp\bake\src\Shell\Task\FixtureTask.php:194
D:\...\Apps\sandbox.local\vendor\cakephp\bake\tests\TestCase\Shell\Task\FixtureTaskTest.php:132
```

This should resolve it. //cc @markstory

Commit:
Provide BC typehinting for TableSchema.

antograssiot reviewed (approved):
:+1:

lorenzo commented:
I'm not too happy that we have to load those classes in every single request

dereuromark commented:
I basically copied the approach of https://github.com/cakephp/cakephp/blob/3.next/src/Http/backwards_compat.php
What's the other option?

markstory commented:
@lorenzo Typehints don't force classes to load, which in turn causes the aliases to emit warnings/errors. Another option would be to not force load the classes but to have the BC aliases be empty classes. e.g.

```php
namespace Cake\Database\Schema;

class Table extends TableSchema {
}
```

chinpei215 commented:
@markstory I don't think inheritance solves this issue. It would cause another strict errors, as people cannot pass the `Table` instances to any methods typehinted as `TableSchema $table`. What if we call `class_alias()` in `TableSchema.php`, and call `class_exists()` in the deprecated `Table.php`?

markstory commented:
@chinpei215 That is worth trying.

markstory committed:
Update backwards compat alias generation.

Generate aliases in the new files. This allows us to resolve issues
where typehints result in fatal errors by ensuring that aliases exist
when the new class is used. The old class location forces the new
location to load which ensures that both names resolve to the same state
(class + alias loaded).

Unfortunately this requires relaxing the PSR1 side-effect rules as the
`@codingStandardsIgnoreStart` tags did not work.

markstory commented:
@chinpei215 I've updated the aliases as you suggested. It works well, but I had to relax the PHPCS rules as the PSR1 side-effect sniff doesn't respect `@codingStandardsIgnore(Start|End)` directives.

ADmad commented on src/Http/Client.php:
```diff
@@ -549,3 +549,5 @@ protected function _createAuth($auth, $options)
     }
 }
+// @deprecated Backwards compatibility with earler 3.x versions.
+class_alias('Cake\Http\Client', 'Cake\Network\Http\Client');
```
Would this and other similar aliases referencing classes in `Cake\Networks` create problem if we want to provide `Cake\Http` as a standalone lib?

markstory replied:
They shouldn't. The aliases just define additional classnames that resolve to the class on the left. It _could_ cause duplicate class name errors if userland code defined classes with the same names as those on the right of the `class_alias()` call.

dereuromark replied:
But userland code should not use the Cake namespace, so we should be fine?

markstory replied:
I agree that userland code shouldn't be defining classes in a namespace they don't own, and that is a reasonable reply to anyone who hits this exotic condition.

ADmad replied:
Okay, cool. Having `Cake\Http` as a standalone lib would be real nice.

markstory replied:
@ADmad I agree. We're getting pretty close to that being possible.

chinpei215 commented:
Looks good to me.
